internal/v4/content_test.go

```diff
@@ -145,6 +138,75 @@ func (s *APISuite) TestServeDiagram(c *gc.C) {
 	})
 }
 
+func (s *APISuite) TestServeDiagramNoPosition(c *gc.C) {
+	patchArchiveCacheAges(s)
+	bundle := &testingBundle{
+		data: &charm.BundleData{
+			Services: map[string]*charm.ServiceSpec{
+				"wordpress": {
+					Charm: "wordpress",
+				},
+				"mysql": {
+					Charm: "utopic/mysql-23",
+					Annotations: map[string]string{
+						"gui-x": "200",
+						"gui-y": "200",
+					},
+				},
+			},
+		},
+	}
+
+	url := newResolvedURL("cs:~charmers/bundle/wordpressbundle-42", 42)
+	err := s.store.AddBundle(bundle, charmstore.AddParams{
+		URL:      url,
+		BlobName: "blobName",
+		BlobHash: fakeBlobHash,
+		BlobSize: fakeBlobSize,
+	})
+	c.Assert(err, gc.IsNil)
+	err = s.store.SetPerms(&url.URL, "read", params.Everyone, url.URL.User)
+	c.Assert(err, gc.IsNil)
+
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     storeURL("bundle/wordpressbundle/diagram.svg"),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK, gc.Commentf("body: %q", rec.Body.Bytes()))
+	c.Assert(rec.Header().Get("Content-Type"), gc.Equals, "image/svg+xml")
+	assertCacheControl(c, rec.Header(), true)
+
+	// Check that the output contains valid XML with an SVG tag,
+	// but don't check the details of the output so that this test doesn't
+	// break every time the jujusvg presentation changes.
+	// Also check that we get an image for each service containing the charm
+	// icon link.
+	assertXMLContains(c, rec.Body.Bytes(), map[string]func(xml.Token) bool{
+		"svg element":    isStartElementWithName("svg"),
+		"wordpress icon": isStartElementWithAttr("image", "href", "../../wordpress/icon.svg"),
+		"mysql icon":     isStartElementWithAttr("image", "href", "../../utopic/mysql-23/icon.svg"),
+	})
+
+	// Do the same check again, but with the short form of the id;
+	// the relative links should change accordingly.
+	rec = httptesting.DoRequest(c, httptesting.DoRequestParams{
+		Handler: s.srv,
+		URL:     storeURL("wordpressbundle/diagram.svg"),
+	})
+	c.Assert(rec.Code, gc.Equals, http.StatusOK, gc.Commentf("body: %q", rec.Body.Bytes()))
+
+	// Check that the output contains valid XML with an SVG tag,
+	// but don't check the details of the output so that this test doesn't
+	// break every time the jujusvg presentation changes.
+	// Also check that we get an image for each service containing the charm
+	// icon link.
+	assertXMLContains(c, rec.Body.Bytes(), map[string]func(xml.Token) bool{
+		"svg element":    isStartElementWithName("svg"),
+		"wordpress icon": isStartElementWithAttr("image", "href", "../wordpress/icon.svg"),
+		"mysql icon":     isStartElementWithAttr("image", "href", "../utopic/mysql-23/icon.svg"),
+	})
```

Comment:
Shouldn't it also check that wordpress charm has annotations for gui-x and gui-y?